Loop through paginated response to check for cluster id

nsAstro opened the pull request:
Loops through the paginated emr list_clusters endpoint and searches each set for the given cluster_id.

closes: #29712

Commit:
Loop through paginated response to check for cluster

thirtyseven commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -68,22 +68,24 @@ def get_cluster_id_by_name(self, emr_cluster_name: str, cluster_states: list[str
-        else:
-            self.log.info("No cluster found for name %s", emr_cluster_name)
-            return None
+            if len(matching_clusters) == 1:
```
This logic isn't quite right, what if two clusters have the same name but are on different pages?

nsAstro replied:
Addressed this in latest commit. I missed that cluster_name was non-unique.

thirtyseven commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -68,22 +68,24 @@ def get_cluster_id_by_name(self, emr_cluster_name: str, cluster_states: list[str
+                return cluster_id
+            elif len(matching_clusters) > 1:
+                raise AirflowException(f"More than one cluster found for name {emr_cluster_name}")
+            elif response["Marker"]:
```
boto has a builtin pagination API, it would be nicer to use it rather than reimplement the logic:
https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

nsAstro replied:
Yeah, I was optimizing for speed when I thought we were filtering on a unique ID. Just rewrote to use the above.

Commit:
Use Boto pagination

Commit:
Remove case for old pagination marker

Commit:
Add test case using Moto to test for 51 clusters list as well as duplicated name

Commit:
Fix test and correct function logic to use page instead of response

Commit:
Fix tests

o-nikolas commented on tests/providers/amazon/aws/hooks/test_emr.py:
```diff
@@ -181,6 +182,11 @@ def test_get_cluster_id_by_name(self):
 
+        for i in range(51):
+            hook.create_job_flow(
+                {"Name": "test_cluster" + str(i % 50), "Instances": {"KeepJobFlowAliveWhenNoSteps": True}}
```
Did you use `i % 50` so that the 0 case would be in the list twice? If so it's a bit sneaky and non-obvious IMHO.

nsAstro replied:
Yes I did. Was just a fast way to test, but I can add a comment to make it clearer or just hardcode a more explicit duplicated cluster name.

o-nikolas replied:
Addressed

o-nikolas commented on tests/providers/amazon/aws/hooks/test_emr.py:
```diff
@@ -191,6 +197,9 @@ def test_get_cluster_id_by_name(self):
 
+        with pytest.raises(AirflowException):
+            hook.get_cluster_id_by_name("test_cluster0", ["RUNNING", "WAITING", "BOOTSTRAPPING"])
+
```
Should you not add a test case which covers the issue this PR is solving? I.e. A success case for getting a cluster id for a cluster that would be on the second page.

nsAstro replied:
I'm not sure if list-clusters API on EMR is ordered at all, so I don't know if there's a way to guarantee that a given test-case cluster shows up on the second page only. If there's a way to guarantee this, I can add that as a test. I'll poke through the AWS docs to see if they provide some sort of ordering on how they list clusters to see if there's a deterministic test case I can write.

nsAstro replied:
Their API [docs](https://docs.aws.amazon.com/emr/latest/APIReference/API_ListClusters.html) says unordered list. I could change the internal logic of the function to order the cluster, but that feels silly to do just to make a test.

vincbeck replied:
I dont think that's what @o-nikolas is saying. I think he wants to have a test to cover the use case **in case** the cluster you are looking for is in the second page. The implementation should not change, just adding a test that covers this use case would increase the testing coverage. And I agree with that

nsAstro replied:
I totally understand, I just am not sure how to force a given clusterName to be on the second page. Can I assume that even if the listClusters API provides an unordered response, it will provide that same order in subsequent queries? Then I can do a regular listClusters and fetch a cluster on the second page. But if the ordering is not consistent, I don't really have a surefire way to guarantee that the clusterName is on the second page. I could increase the number of clusters (<1000) to make the probability very high.

I'm happy to add tests, even for testing duplicate clusterNames which wasn't part of this PR. I'm just not too familiar with the EMR API, so I'm looking for some guidance.

nsAstro replied:
I just realized with the way the marker works, that listClusters will be consistent. I'll edit the test accordingly.

o-nikolas replied:
Addressed

Commit:
Fix tests and force case for second page EMR cluster

Commit:
Formatting

vincbeck commented on tests/providers/amazon/aws/hooks/test_emr.py:
```diff
@@ -191,6 +201,24 @@ def test_get_cluster_id_by_name(self):
 
         assert no_match is None
 
+        with pytest.raises(AirflowException):
```
LGTM but I would move it to a separate test for clarity purposes. The shorter the tests are, the more likely people will read them, the better :)

o-nikolas replied:
+1

o-nikolas replied:
Addressed

uranusjr commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -69,11 +69,14 @@ def get_cluster_id_by_name(self, emr_cluster_name: str, cluster_states: list[str
+        for page in response_iterator:
+            matching_clusters.extend(
+                list(filter(lambda cluster: cluster["Name"] == emr_cluster_name, page["Clusters"]))
+            )
```
```suggestion
        matching_clusters = [
            cluster
            for page in response_iterator
            for cluster in page["Clusters"]
            if cluster["Name"] == emr_cluster_name
        ]
```

uranusjr commented on tests/providers/amazon/aws/hooks/test_emr.py:
```diff
@@ -174,13 +175,22 @@ def test_emr_connection(self, aws_conn_id, emr_conn_id):
         """
         Test that we can resolve cluster id by cluster name.
+        Create 50 clusters with the range 1-50 appended and create a duplicate of index 1
+        to test logic.
```
```suggestion

        Create 50 clusters with the range 1-50 appended and create a duplicate of index 1
        to test logic.
```

nsAstro committed:
Merge lines to improve readability

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

o-nikolas committed:
Update tests/providers/amazon/aws/hooks/test_emr.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

o-nikolas committed:
Address PR Feedback

- Separate tests into individual methods
- Ensure pagination test checks for cluster id on second page

o-nikolas committed:
Merge branch 'main' into emr_paginate

o-nikolas committed:
Merge branch 'main' into emr_paginate

o-nikolas committed:
Merge branch 'main' into emr_paginate

o-nikolas committed:
Merge branch 'main' into emr_paginate

o-nikolas committed:
Update test_emr.py

Fix indent
